# **Procesamiento de Lenguaje Natural**

## Maestría en Inteligencia Artificial Aplicada
#### Tecnológico de Monterrey
#### Prof Luis Eduardo Falcón Morales

### **Adtividad en Equipos: sistema LLM + RAG**

* **Nombres, matrículas y correos:**

  * José Emiliano Riosmena Castañón — A01704245 — A01704245@tec.mx
  * Sofía Donlucas Bañuelos — A01655565 — A01655565@tec.mx
  * Sebastián Estrada García — A00805402 — A00805402@tec.mx
  * Gerardo González Martínez — A01840096 — A01840096@tec.mx

* **Número de Equipo:** 3
* **Fecha de entrega:** 14 de junio de 2026
* **Repositorio del proyecto:** https://github.com/gerardoglezmart/chatbot-rag-iayaa
* **Demo en línea (Hugging Face Spaces):** https://huggingface.co/spaces/GerardoTec/chatbot-rag-iayaa


## 1. Introducción y problemática

El curso de Inteligencia Artificial y Aprendizaje Automático (`IAyAA`) cubre conceptos, algoritmos y métricas fundamentales que suelen consultarse de manera recurrente durante el estudio y la resolución de actividades. Sin embargo, ese contenido se encuentra distribuido entre material teórico en PDF, sesiones y documentos complementarios, lo que vuelve más lenta la localización de definiciones, comparaciones entre modelos y explicaciones puntuales sobre temas como gradiente descendente, entropía, sesgo-varianza o máquinas de soporte vectorial. En este contexto, un chatbot académico con arquitectura `RAG + LLM` puede centralizar el acceso al contenido del curso y responder preguntas apoyándose en evidencia documental recuperada desde el corpus local.

El usuario objetivo de este sistema es un estudiante del curso `IAyAA` que necesita consultar rápidamente explicaciones conceptuales, diferencias entre técnicas o recordatorios sobre algoritmos vistos en clase. El alcance de esta primera versión se limita a responder en español preguntas académicas sobre el subconjunto temático seleccionado del curso, utilizando únicamente documentos concentrados en `04_datos_rag/`. Por diseño, el sistema no pretende reemplazar materiales oficiales, resolver ejercicios completos fuera del contexto recuperado ni responder sobre temas no incluidos en el corpus; su función principal es apoyar el estudio con respuestas contextualizadas, breves y trazables a fuentes del curso.

**Alcance operativo de esta versión:**
- Responder preguntas conceptuales sobre temas del corpus `IAyAA`.
- Recuperar fragmentos relevantes antes de generar la respuesta.
- Mostrar fuentes recuperadas para hacer visible la trazabilidad.
- Funcionar dentro del notebook como una demo académica reproducible.

**Límites del sistema:**
- No cubre todo el repositorio ni todos los materiales del curso.
- Depende de la calidad de las transcripciones y de la selección de fragmentos recuperados.
- Puede dar respuestas parciales si el contexto recuperado es insuficiente.
- No sustituye una explicación docente ni una validación académica formal.


## 2. Justificación de RAG + LLM

El problema que se busca resolver en este proyecto no es únicamente generar texto sobre inteligencia artificial, sino responder preguntas académicas del curso `IAyAA` con apego a los materiales reales de estudio. Un `LLM` usado de forma aislada puede producir respuestas fluidas y plausibles, pero también corre el riesgo de mezclar conceptos, omitir matices vistos en clase o responder de manera demasiado general. En un contexto académico esto es especialmente delicado, porque el estudiante necesita explicaciones consistentes con el contenido del curso y no solo definiciones amplias tomadas del conocimiento general del modelo.

La arquitectura `RAG` permite reducir ese problema al recuperar primero fragmentos relevantes del corpus documental y usar después esos fragmentos como contexto para la generación. En este caso, la recuperación se hace sobre transcripciones y documentos teóricos seleccionados del curso, lo que vuelve más probable que la respuesta incluya terminología, ejemplos y enfoques alineados con `IAyAA`. Así, el sistema no depende únicamente de la memoria paramétrica del modelo, sino de evidencia textual concreta extraída del dominio que se desea consultar.

La combinación `RAG + LLM` también se justifica porque equilibra precisión y claridad expositiva. El componente de recuperación aporta pertinencia temática y trazabilidad documental, mientras que el `LLM` transforma el contexto recuperado en una respuesta integrada, legible y útil para el usuario final. Esto resulta adecuado para un chatbot académico en español, ya que permite contestar preguntas conceptuales, comparativas y explicativas con base documental, sin requerir reentrenar el modelo ni construir una infraestructura compleja para la primera versión del proyecto.

## 3. Selección del LLM

Para esta primera versión del chatbot se eligió trabajar con OpenAI y usar `gpt-5-mini` como configuración base del componente generativo. La decisión no busca maximizar sofisticación, sino mantener un equilibrio adecuado entre disponibilidad, facilidad de integración y calidad suficiente para una entrega académica en notebook. Dado que el proyecto requiere responder en español con apoyo de contexto recuperado, interesa un modelo capaz de seguir instrucciones con precisión, sintetizar información técnica y operar de forma estable dentro de un flujo de prompting contextual.

El criterio de selección prioriza cuatro factores. Primero, la facilidad de integración en Python mediante el cliente oficial de OpenAI, lo que simplifica la conexión entre recuperación, construcción de prompt y generación final. Segundo, la disponibilidad práctica en el entorno de trabajo, ya que la integración puede activarse de forma controlada dentro del notebook y permite trabajar con credenciales sin exponerlas en el contenido visible del entregable. Tercero, el costo operativo, porque `gpt-5-mini` ofrece un perfil razonable para pruebas académicas y demostraciones iterativas sin exigir una configuración pesada. Cuarto, el desempeño esperado en comprensión y redacción en español, suficiente para transformar el contexto recuperado en respuestas claras para un estudiante de maestría.

También se mantuvo una decisión de diseño importante: desacoplar la interfaz generativa del resto del sistema. Aunque la configuración actual usa OpenAI con `gpt-5-mini`, el notebook conserva una estructura basada en una función tipo `generate_answer(contexto, pregunta)` para que el proveedor pueda sustituirse sin rehacer el pipeline RAG. Esta elección vuelve más robusto el entregable, porque permite justificar el modelo seleccionado hoy y, al mismo tiempo, conservar flexibilidad técnica si en otra ejecución se requiere cambiar de modelo, ajustar costos o trabajar temporalmente con un modo de respaldo extractivo.


**Alternativas consideradas (abiertas y propietarias):**

| Modelo | Tipo | Ventajas | Rol en este proyecto |
|---|---|---|---|
| `gpt-5-mini` (OpenAI) | Propietario | Excelente español, API estable, costo bajo por consulta, seguimiento fino de instrucciones | **Backend principal** (`openai`) |
| `Qwen2.5-7B-Instruct` (Alibaba) | Código abierto | Buen desempeño en español, licencia permisiva, disponible vía Hugging Face Inference API | Backend `hf`: alternativa abierta en la nube, usada en el despliegue del Space |
| `llama3.2:3b` (Meta) | Código abierto | Corre 100% local en CPU con 8 GB de RAM vía Ollama; sin costo por token ni dependencia de red | Backend `ollama`: alternativa abierta local |
| Respaldo extractivo | — | Sin dependencias externas, siempre disponible | Backend `extractivo`: degradación elegante |

A diferencia de una justificación puramente teórica, en este proyecto la comparación entre modelos
abiertos y propietarios está **implementada**: la sección 5.5 define los cuatro backends como funciones
intercambiables y la variable `LLM_BACKEND` (o el selector de la interfaz gráfica) permite elegir entre
ellos sin modificar el pipeline RAG. Esto demuestra empíricamente la decisión de diseño de desacoplar
la generación de la recuperación: el mismo contexto recuperado puede ser respondido por un modelo
propietario (`gpt-5-mini`), por un modelo abierto en la nube (`Qwen2.5-7B`) o por un modelo abierto
ejecutándose localmente (`llama3.2:3b`).


## 4. Corpus documental

El corpus inicial se concentró en `04_datos_rag/` con un criterio de cobertura temática controlada y facilidad de procesamiento para una primera versión del sistema. En lugar de incluir todo el curso, se seleccionaron `8` documentos que cubren conceptos nucleares de `IAyAA` y permiten probar preguntas introductorias, comparativas e integradoras.

**Archivos incluidos en el corpus:**
- `01_introduccion_ia_teoria.txt`
- `02_gradiente_descendente.txt`
- `03_regresion_logistica.txt`
- `04_entropia_ganancia_informacion.txt`
- `05_sesgo_varianza_curvas_aprendizaje.txt`
- `06_svm_kernel_holgura.txt`
- `07_introduccion_ia_teoria.pdf`
- `08_metricas_clasificacion_matriz_confusion.pdf`

**Cobertura temática principal:**
- introducción a la inteligencia artificial y al aprendizaje automático,
- gradiente descendente y optimización,
- regresión logística,
- entropía y ganancia de información,
- sesgo-varianza y curvas de aprendizaje,
- `SVM`, kernel radial y variables de holgura,
- métricas de clasificación y matriz de confusión.

La decisión de priorizar `6` transcripciones `.txt` responde a una razón práctica: son más limpias para carga, limpieza y chunking dentro del notebook. Los `2` archivos `.pdf` se conservaron como respaldo teórico y como evidencia de que el pipeline puede trabajar con más de un formato documental sin salir del alcance de la actividad.

En conjunto, este corpus ofrece tres ventajas para la entrega. Primero, mantiene un dominio coherente y directamente alineado con la materia. Segundo, evita ruido innecesario al no mezclar temas periféricos todavía no evaluados. Tercero, permite explicar con claridad la trazabilidad de las respuestas, porque cada consulta puede vincularse con fuentes concretas del curso.


In [1]:
# 5.1 Configuración inicial y librerías
# El notebook corre igual en local (conda) que en Google Colab:
# - en Colab instala dependencias y, si hace falta, monta Drive;
# - en local usa la carpeta relativa 04_datos_rag del repositorio.

import os
import re
import sys
import unicodedata
from pathlib import Path

try:
    import google.colab  # noqa: F401
    IN_COLAB = True
except ImportError:
    IN_COLAB = False

if IN_COLAB:
    %pip install -q pypdf scikit-learn pandas openai ipywidgets sentence-transformers gradio huggingface_hub

import numpy as np
import pandas as pd
from IPython.display import display
from pypdf import PdfReader
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

_CANDIDATAS = [
    Path("04_datos_rag"),
    Path("/content/drive/MyDrive/Colab Notebooks NLP/Actividad 5/04_datos_rag"),
]
DATA_DIR = next((p for p in _CANDIDATAS if p.exists()), None)
if DATA_DIR is None and IN_COLAB:
    from google.colab import drive
    drive.mount("/content/drive")
    DATA_DIR = _CANDIDATAS[1] if _CANDIDATAS[1].exists() else None
assert DATA_DIR is not None, "No se encontró la carpeta 04_datos_rag"
print(f"Entorno: {'Colab' if IN_COLAB else 'local'} | Corpus: {DATA_DIR.resolve()}")

TOPIC_MAP = {
    "01_introduccion_ia_teoria": "Introducción a IA y aprendizaje automático",
    "02_gradiente_descendente": "Regresión y gradiente descendente",
    "03_regresion_logistica": "Clasificación y regresión logística",
    "04_entropia_ganancia_informacion": "Árboles de decisión, entropía y ganancia de información",
    "05_sesgo_varianza_curvas_aprendizaje": "Sesgo-varianza y curvas de aprendizaje",
    "06_svm_kernel_holgura": "SVM, kernel radial y variables de holgura",
    "07_introduccion_ia_teoria": "Introducción a IA y aprendizaje automático",
    "08_metricas_clasificacion_matriz_confusion": "Métricas de clasificación y matriz de confusión",
}


Entorno: local | Corpus: /home/nemis/Documents/MNA/TC5060.10/Actividad-05_Semanas-07-08_Chatbot-RAG-LLM/04_datos_rag


## 5. Implementación del sistema RAG

Esta sección concentra el desarrollo técnico del chatbot y describe cómo se articula el flujo completo desde la lectura del corpus hasta la generación de la respuesta final. La arquitectura implementada sigue un esquema `RAG` demostrable dentro del notebook, priorizando claridad, estabilidad y trazabilidad sobre complejidad de infraestructura.

### Arquitectura del sistema

El siguiente diagrama resume el flujo completo, desde la ingesta del corpus hasta la generación de la respuesta con fuentes citadas:

![Flujo del sistema RAG + LLM](capturas/diagrama/diagrama1.png)


El primer bloque del sistema corresponde a la **carga de documentos**. Para ello se leen archivos `.txt` y `.pdf` desde `04_datos_rag/` y se normalizan a una estructura común con campos como `source`, `text`, `tipo` y `tema`. Esta homogeneización permite tratar de forma consistente documentos provenientes de transcripciones y materiales teóricos, y además deja preparada la base para auditar qué fuente alimenta cada respuesta.

El segundo bloque es la **preparación del texto y el chunking**. Después de una limpieza ligera de espacios, saltos de línea y fragmentos ruidosos, cada documento se divide en bloques manejables (900 caracteres con traslape de 150). El propósito del chunking es conservar suficiente contexto local sin obligar al sistema a comparar preguntas contra documentos completos demasiado largos. Así, cada chunk se vuelve la unidad básica de recuperación semántica y de trazabilidad dentro del notebook.

El tercer bloque cubre la **representación vectorial y la recuperación semántica**. Los chunks se codifican con el modelo de embeddings neuronales multilingüe `paraphrase-multilingual-MiniLM-L12-v2` (`sentence-transformers`), que proyecta cada fragmento y cada consulta a un espacio vectorial denso donde la cercanía refleja similitud de significado y no solo coincidencia de palabras. La consulta del usuario se codifica en el mismo espacio y se rankean los fragmentos por similitud coseno para recuperar un conjunto `top-k`. Como evidencia del beneficio de este enfoque, se conserva el recuperador `TF-IDF` de la primera versión como *baseline* comparativo y la sección 8 cuantifica la diferencia entre ambos.

El cuarto bloque corresponde a la **construcción del prompt y la generación de respuesta**. Los fragmentos recuperados se formatean como contexto explícito y se incorporan a un prompt que obliga al modelo a responder en español, limitarse al material recuperado y declarar insuficiencia si la evidencia no basta. El componente generativo es **intercambiable**: `gpt-5-mini` (propietario, OpenAI), `claude-opus-4-8` (propietario, Anthropic), `Qwen2.5-7B-Instruct` (abierto, vía Hugging Face Inference API), `llama3.2:3b` (abierto, ejecución local con Ollama) o un modo extractivo de respaldo. De esta manera, el sistema conserva continuidad funcional incluso cuando ningún `LLM` está disponible.

Las **librerías principales** utilizadas en esta implementación responden a un criterio de simplicidad. `pandas` organiza documentos, chunks y resultados en estructuras tabulares fáciles de inspeccionar; `pypdf` extrae texto de los materiales teóricos en PDF; `sentence-transformers` aporta el modelo de embeddings semánticos multilingües; `scikit-learn` provee `TfidfVectorizer` y `cosine_similarity` para el baseline léxico; `openai` y `huggingface_hub` conectan con los modelos generativos propietario y abierto; `gradio` construye la interfaz de chat; e `IPython.display` y módulos estándar (`pathlib`, `re`, `os`, `unicodedata`) apoyan visualización, rutas y limpieza.

### Cómo usar el chatbot

1. **Ejecutar las celdas en orden** (en Colab: *Entorno de ejecución → Ejecutar todo*; en local: kernel del entorno `rag-chatbot`).
2. **Configurar el modelo generativo** (sección 5.5). El sistema detecta automáticamente el primer backend disponible:
   - `OPENAI_API_KEY` definida → usa `gpt-5-mini` (en Colab la clave se solicita con `getpass`, sin quedar guardada en el notebook);
   - `HF_TOKEN` definida → usa `Qwen2.5-7B-Instruct` (abierto);
   - servidor Ollama local activo → usa `llama3.2:3b` (abierto);
   - sin nada de lo anterior → modo extractivo local.
3. **Conversar**: la sección 6.2 abre la interfaz de chat (Gradio); se puede escribir cualquier pregunta sobre los temas del corpus o usar los ejemplos precargados. La misma interfaz está desplegada de forma permanente en Hugging Face Spaces (liga al inicio del notebook).
4. **Verificar la trazabilidad**: cada respuesta incluye las fuentes recuperadas con su puntaje de similitud.

Finalmente, el sistema conserva **límites conocidos** que deben reconocerse en la revisión académica. La calidad de la recuperación depende del ruido presente en algunas transcripciones y de la cobertura del corpus seleccionado. La lectura de PDFs puede heredar problemas de formato y la generación con modelos externos depende de credenciales y cuota; por ello se mantienen el backend local abierto y el respaldo extractivo.

En conjunto, el flujo implementado puede resumirse en la siguiente secuencia:

1. Carga del corpus.
2. Limpieza y preparación del texto.
3. Chunking.
4. Embeddings semánticos del corpus.
5. Recuperación semántica (top-k por similitud coseno).
6. Construcción del prompt.
7. Generación de respuesta (backend intercambiable).
8. Presentación de fuentes recuperadas.


In [2]:
# 5.2 Carga y extracción de documentos
# Lectura de archivos .txt y .pdf con una estructura homogénea.

def infer_topic(path: Path) -> str:
    return TOPIC_MAP.get(path.stem, "Tema por clasificar")


def read_txt(path: Path) -> str:
    return path.read_text(encoding="utf-8", errors="ignore")


def read_pdf(path: Path) -> str:
    reader = PdfReader(str(path))
    pages = [(page.extract_text() or "") for page in reader.pages]
    return "\n".join(pages)


def load_document(path: Path) -> dict:
    suffix = path.suffix.lower()
    if suffix == ".txt":
        text = read_txt(path)
    elif suffix == ".pdf":
        text = read_pdf(path)
    else:
        raise ValueError(f"Tipo de archivo no soportado: {path.name}")

    return {
        "source": path.name,
        "text": text,
        "tipo": suffix.lstrip("."),
        "tema": infer_topic(path),
    }


def load_corpus(data_dir: Path = DATA_DIR, allowed_suffixes=(".txt", ".pdf")) -> pd.DataFrame:
    documents = []
    for path in sorted(data_dir.iterdir()):
        if path.is_file() and path.suffix.lower() in allowed_suffixes:
            documents.append(load_document(path))
    return pd.DataFrame(documents)


documents_df = load_corpus()
documents_df["num_chars"] = documents_df["text"].str.len()
documents_df["num_words"] = documents_df["text"].str.split().str.len()

display(documents_df[["source", "tipo", "tema", "num_chars", "num_words"]])
print(f"Documentos cargados: {len(documents_df)}")

,source,tipo,tema,num_chars,num_words
0,01_introduccion_ia_teoria.txt,txt,Introducción a IA y aprendizaje automático,32313,5684
1,02_gradiente_descendente.txt,txt,Regresión y gradiente descendente,9499,1652
2,03_regresion_logistica.txt,txt,Clasificación y regresión logística,21792,3937
3,04_entropia_ganancia_informacion.txt,txt,"Árboles de decisión, entropía y ganancia de in...",27562,4980
4,05_sesgo_varianza_curvas_aprendizaje.txt,txt,Sesgo-varianza y curvas de aprendizaje,20917,3785
5,06_svm_kernel_holgura.txt,txt,"SVM, kernel radial y variables de holgura",30093,5220
6,07_introduccion_ia_teoria.pdf,pdf,Introducción a IA y aprendizaje automático,5010,690
7,08_metricas_clasificacion_matriz_confusion.pdf,pdf,Métricas de clasificación y matriz de confusión,9200,1161


Documentos cargados: 8


In [3]:
# 5.3 Limpieza y chunking
# Preparar el corpus y fragmentarlo
# en bloques manejables para retrieval semántico.

def clean_text(text: str) -> str:
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    text = re.sub(r"[\t\f\v]+", " ", text)
    text = re.sub(r" ?\n ?", "\n", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    text = re.sub(r" {2,}", " ", text)
    return text.strip()


def prepare_corpus(documents: pd.DataFrame, min_chars: int = 80) -> pd.DataFrame:
    prepared = documents.copy()
    prepared["text_raw"] = prepared["text"]
    prepared["text"] = prepared["text"].fillna("").map(clean_text)
    prepared["num_chars_clean"] = prepared["text"].str.len()
    prepared["num_words_clean"] = prepared["text"].str.split().str.len()
    prepared = prepared[prepared["num_chars_clean"] >= min_chars].reset_index(drop=True)
    return prepared


prepared_df = prepare_corpus(documents_df)

display(
    prepared_df[["source", "tipo", "tema", "num_chars", "num_chars_clean", "num_words_clean"]]
)

empty_docs = int((prepared_df["text"].str.strip() == "").sum())
print(f"Documentos listos para chunking: {len(prepared_df)}")
print(f"Documentos vacíos tras limpieza: {empty_docs}")

prepared_df[["source", "text"]].head(2)

CHUNK_SIZE = 900
CHUNK_OVERLAP = 150


def split_into_paragraphs(text: str) -> list[str]:
    paragraphs = [paragraph.strip() for paragraph in re.split(r"\n{2,}", text) if paragraph.strip()]
    return paragraphs or ([text.strip()] if text.strip() else [])


def split_long_paragraph(paragraph: str, chunk_size: int) -> list[str]:
    words = paragraph.split()
    windows = []
    current_words = []

    for word in words:
        candidate = " ".join(current_words + [word]).strip()
        if current_words and len(candidate) > chunk_size:
            windows.append(" ".join(current_words).strip())
            current_words = [word]
        else:
            current_words.append(word)

    if current_words:
        windows.append(" ".join(current_words).strip())

    return [window for window in windows if window]


def chunk_text(text: str, chunk_size: int = CHUNK_SIZE, chunk_overlap: int = CHUNK_OVERLAP) -> list[str]:
    paragraphs = split_into_paragraphs(text)
    chunks = []
    current_parts = []

    for paragraph in paragraphs:
        if len(paragraph) > chunk_size:
            if current_parts:
                chunks.append("\n\n".join(current_parts).strip())
                current_parts = []
            chunks.extend(split_long_paragraph(paragraph, chunk_size))
            continue

        candidate_parts = current_parts + [paragraph]
        candidate_text = "\n\n".join(candidate_parts).strip()

        if current_parts and len(candidate_text) > chunk_size:
            chunks.append("\n\n".join(current_parts).strip())
            if chunk_overlap > 0:
                overlap_text = chunks[-1][-chunk_overlap:].strip()
                current_parts = [overlap_text, paragraph] if overlap_text else [paragraph]
            else:
                current_parts = [paragraph]
        else:
            current_parts = candidate_parts

    if current_parts:
        chunks.append("\n\n".join(current_parts).strip())

    return [chunk for chunk in chunks if chunk]


def build_chunks(corpus: pd.DataFrame, chunk_size: int = CHUNK_SIZE, chunk_overlap: int = CHUNK_OVERLAP) -> pd.DataFrame:
    rows = []

    for _, row in corpus.iterrows():
        text_chunks = chunk_text(row["text"], chunk_size=chunk_size, chunk_overlap=chunk_overlap)
        for chunk_index, chunk in enumerate(text_chunks, start=1):
            rows.append(
                {
                    "chunk_id": f"{Path(row['source']).stem}_chunk_{chunk_index:03d}",
                    "source": row["source"],
                    "tipo": row["tipo"],
                    "tema": row["tema"],
                    "chunk_index": chunk_index,
                    "text": chunk,
                    "num_chars_chunk": len(chunk),
                    "num_words_chunk": len(chunk.split()),
                }
            )

    return pd.DataFrame(rows)


chunks_df = build_chunks(prepared_df)

display(chunks_df[["chunk_id", "source", "tema", "num_chars_chunk", "num_words_chunk"]].head(10))
display(chunks_df.groupby("source").size().reset_index(name="num_chunks"))
print(f"Chunks generados: {len(chunks_df)}")
print(f"Tamaño configurado por chunk: {CHUNK_SIZE} caracteres")
print(f"Traslape configurado: {CHUNK_OVERLAP} caracteres")


,source,tipo,tema,num_chars,num_chars_clean,num_words_clean
0,01_introduccion_ia_teoria.txt,txt,Introducción a IA y aprendizaje automático,32313,32312,5684
1,02_gradiente_descendente.txt,txt,Regresión y gradiente descendente,9499,9498,1652
2,03_regresion_logistica.txt,txt,Clasificación y regresión logística,21792,21791,3937
3,04_entropia_ganancia_informacion.txt,txt,"Árboles de decisión, entropía y ganancia de in...",27562,27561,4980
4,05_sesgo_varianza_curvas_aprendizaje.txt,txt,Sesgo-varianza y curvas de aprendizaje,20917,20916,3785
5,06_svm_kernel_holgura.txt,txt,"SVM, kernel radial y variables de holgura",30093,30092,5220
6,07_introduccion_ia_teoria.pdf,pdf,Introducción a IA y aprendizaje automático,5010,4952,690
7,08_metricas_clasificacion_matriz_confusion.pdf,pdf,Métricas de clasificación y matriz de confusión,9200,9193,1161


Documentos listos para chunking: 8
Documentos vacíos tras limpieza: 0


,chunk_id,source,tema,num_chars_chunk,num_words_chunk
0,01_introduccion_ia_teoria_chunk_001,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,894,151
1,01_introduccion_ia_teoria_chunk_002,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,899,151
2,01_introduccion_ia_teoria_chunk_003,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,898,147
3,01_introduccion_ia_teoria_chunk_004,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,900,143
4,01_introduccion_ia_teoria_chunk_005,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,900,153
5,01_introduccion_ia_teoria_chunk_006,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,893,149
6,01_introduccion_ia_teoria_chunk_007,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,893,142
7,01_introduccion_ia_teoria_chunk_008,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,899,150
8,01_introduccion_ia_teoria_chunk_009,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,897,156
9,01_introduccion_ia_teoria_chunk_010,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,895,165


,source,num_chunks
0,01_introduccion_ia_teoria.txt,36
1,02_gradiente_descendente.txt,11
2,03_regresion_logistica.txt,25
3,04_entropia_ganancia_informacion.txt,31
4,05_sesgo_varianza_curvas_aprendizaje.txt,24
5,06_svm_kernel_holgura.txt,34
6,07_introduccion_ia_teoria.pdf,6
7,08_metricas_clasificacion_matriz_confusion.pdf,11


Chunks generados: 178
Tamaño configurado por chunk: 900 caracteres
Traslape configurado: 150 caracteres


In [4]:
# 5.4 Embeddings y recuperación semántica
# Recuperador principal: embeddings neuronales multilingües (sentence-transformers).
# La primera versión del sistema usaba TF-IDF con expansión manual de consultas;
# se conserva aquí como baseline comparativo (la sección 8 cuantifica la mejora).

from sentence_transformers import SentenceTransformer

EMBEDDING_MODEL_NAME = "paraphrase-multilingual-MiniLM-L12-v2"

st_model = SentenceTransformer(EMBEDDING_MODEL_NAME)
corpus_embeddings = st_model.encode(
    chunks_df["text"].tolist(),
    normalize_embeddings=True,
    show_progress_bar=True,
)

chunk_embeddings_df = chunks_df.copy()
chunk_embeddings_df["embedding_index"] = range(len(chunk_embeddings_df))
chunk_embeddings_df["embedding_model"] = EMBEDDING_MODEL_NAME

print(f"Modelo de embeddings activo: {EMBEDDING_MODEL_NAME}")
print(f"Matriz de embeddings del corpus: {corpus_embeddings.shape}")


def recuperar_contexto(pregunta: str, k: int = 5) -> pd.DataFrame:
    """Recuperación semántica: similitud coseno sobre embeddings normalizados."""
    query_embedding = st_model.encode([pregunta], normalize_embeddings=True)
    similarities = (corpus_embeddings @ query_embedding.T).flatten()
    top_indices = similarities.argsort()[::-1][:k]

    results = chunk_embeddings_df.iloc[top_indices].copy()
    results["score_similitud"] = similarities[top_indices]
    results["pregunta"] = pregunta
    return results[
        ["pregunta", "chunk_id", "source", "tema", "chunk_index", "score_similitud", "text"]
    ].reset_index(drop=True)


# --- Baseline TF-IDF (configuración de la primera versión, para comparación) ---

tfidf_vectorizer = TfidfVectorizer(
    lowercase=True,
    strip_accents="unicode",
    ngram_range=(1, 2),
    min_df=1,
    max_df=0.95,
    token_pattern=r"(?u)\b[a-záéíóúüñ]{2,}\b",
    sublinear_tf=True,
)
tfidf_embeddings = tfidf_vectorizer.fit_transform(chunks_df["text"])


def recuperar_contexto_tfidf(pregunta: str, k: int = 5) -> pd.DataFrame:
    """Baseline léxico: similitud coseno sobre vectores TF-IDF."""
    query_vec = tfidf_vectorizer.transform([pregunta])
    similarities = cosine_similarity(query_vec, tfidf_embeddings).flatten()
    top_indices = similarities.argsort()[::-1][:k]

    results = chunk_embeddings_df.iloc[top_indices].copy()
    results["score_similitud"] = similarities[top_indices]
    results["pregunta"] = pregunta
    return results[
        ["pregunta", "chunk_id", "source", "tema", "chunk_index", "score_similitud", "text"]
    ].reset_index(drop=True)


pregunta_prueba = "¿Qué es el gradiente descendente y para qué sirve en aprendizaje automático?"
print("\nTop-3 semántico:")
display(recuperar_contexto(pregunta_prueba, k=3)[["chunk_id", "source", "tema", "score_similitud"]])
print("Top-3 TF-IDF (baseline):")
display(recuperar_contexto_tfidf(pregunta_prueba, k=3)[["chunk_id", "source", "tema", "score_similitud"]])


/home/nemis/anaconda3/envs/rag-chatbot/lib/python3.11/site-packages/torch/cuda/__init__.py:187: UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 12040). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at /pytorch/c10/cuda/CUDAFunctions.cpp:119.)
  return torch._C._cuda_getDeviceCount() > 0


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

Batches:   0%|          | 0/6 [00:00<?, ?it/s]

Modelo de embeddings activo: paraphrase-multilingual-MiniLM-L12-v2
Matriz de embeddings del corpus: (178, 384)

Top-3 semántico:


,chunk_id,source,tema,score_similitud
0,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.565466
1,02_gradiente_descendente_chunk_007,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.557666
2,02_gradiente_descendente_chunk_006,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.532420


Top-3 TF-IDF (baseline):


,chunk_id,source,tema,score_similitud
0,02_gradiente_descendente_chunk_007,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.136475
1,02_gradiente_descendente_chunk_001,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.101255
2,07_introduccion_ia_teoria_chunk_001,07_introduccion_ia_teoria.pdf,Introducción a IA y aprendizaje automático,0.093816


In [5]:
# 5.5 Prompt y componente generativo intercambiable
# Cuatro backends: OpenAI (propietario), Hugging Face y Ollama (código abierto)
# y un respaldo extractivo local. Sin credenciales expuestas en el notebook.

PROMPT_BASE = """Rol: Eres un asistente académico del curso Inteligencia Artificial y Aprendizaje Automático (IAyAA).

Objetivo:
- Responder preguntas del curso apoyándote en el contexto recuperado del material oficial.

Instrucciones obligatorias:
- Responde siempre en español.
- Basa tu respuesta en la información del contexto recuperado; puedes reorganizarla y sintetizarla con tus propias palabras.
- Si el contexto cubre el tema solo de forma parcial, responde con lo que sí está presente y aclara en una frase qué aspecto no quedó cubierto.
- Únicamente si el contexto no contiene nada relevante para la pregunta, inicia con: 'Con el contexto recuperado no es posible responder con suficiente precisión'.
- No inventes definiciones, fórmulas o ejemplos ajenos al contexto.
- Tono claro y preciso para un estudiante de maestría; menciona de forma breve la fuente o tema del que sale la respuesta.

Formato de salida esperado:
- Respuesta de 3 a 6 oraciones.
"""

LLM_BACKEND = os.environ.get("LLM_BACKEND", "auto")  # auto | openai | claude | hf | ollama | extractivo
OPENAI_MODEL = os.environ.get("OPENAI_MODEL", "gpt-5-mini")
OPENAI_REASONING_EFFORT = os.environ.get("OPENAI_REASONING_EFFORT", "low")
ANTHROPIC_MODEL = os.environ.get("ANTHROPIC_MODEL", "claude-opus-4-8")
HF_MODEL = os.environ.get("HF_MODEL", "Qwen/Qwen2.5-7B-Instruct")
OLLAMA_MODEL = os.environ.get("OLLAMA_MODEL", "llama3.2:3b")
OLLAMA_BASE_URL = os.environ.get("OLLAMA_BASE_URL", "http://localhost:11434/v1")


def format_contexto(contexto_df: pd.DataFrame) -> str:
    bloques = []
    for row in contexto_df.itertuples(index=False):
        bloques.append(
            f"[Fuente: {row.source} | Chunk: {row.chunk_id} | Score: {row.score_similitud:.4f}]\n{row.text}"
        )
    return "\n\n".join(bloques)


def construir_prompt(pregunta: str, contexto_df: pd.DataFrame) -> str:
    contexto = format_contexto(contexto_df)
    return (
        f"{PROMPT_BASE}\n"
        "Contexto recuperado:\n"
        "--------------------\n"
        f"{contexto}\n\n"
        "Pregunta del usuario:\n"
        "---------------------\n"
        f"{pregunta}\n\n"
        "Respuesta final:\n"
    )


def get_openai_llm_callable():
    """Backend propietario: gpt-5-mini vía API de OpenAI."""
    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key and IN_COLAB:
        from getpass import getpass
        api_key = getpass("OPENAI_API_KEY (Enter para omitir): ").strip()
        if api_key:
            os.environ["OPENAI_API_KEY"] = api_key
    if not api_key:
        return None

    from openai import OpenAI
    client = OpenAI(api_key=api_key)

    def llm_openai(prompt: str) -> str:
        response = client.responses.create(
            model=OPENAI_MODEL,
            input=prompt,
            reasoning={"effort": OPENAI_REASONING_EFFORT},
            store=False,
        )
        return response.output_text.strip()

    return llm_openai


def get_claude_llm_callable():
    """Backend propietario alternativo: Claude vía el SDK oficial de Anthropic."""
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        return None

    import anthropic
    client = anthropic.Anthropic(api_key=api_key)

    def llm_claude(prompt: str) -> str:
        response = client.messages.create(
            model=ANTHROPIC_MODEL,
            max_tokens=1024,
            messages=[{"role": "user", "content": prompt}],
        )
        return next(
            (block.text for block in response.content if block.type == "text"), ""
        ).strip()

    return llm_claude


def get_hf_llm_callable():
    """Backend de código abierto en la nube: Qwen2.5-7B vía HF Inference API."""
    token = os.environ.get("HF_TOKEN")
    if not token:
        return None

    from huggingface_hub import InferenceClient
    client = InferenceClient(model=HF_MODEL, token=token)

    def llm_hf(prompt: str) -> str:
        response = client.chat_completion(
            messages=[{"role": "user", "content": prompt}],
            max_tokens=500,
            temperature=0.2,
        )
        return response.choices[0].message.content.strip()

    return llm_hf


def get_ollama_llm_callable():
    """Backend de código abierto local: llama3.2:3b servido por Ollama."""
    import urllib.request
    try:
        urllib.request.urlopen(OLLAMA_BASE_URL.removesuffix("/v1"), timeout=2)
    except Exception:
        return None

    from openai import OpenAI
    client = OpenAI(base_url=OLLAMA_BASE_URL, api_key="ollama")

    def llm_ollama(prompt: str) -> str:
        response = client.chat.completions.create(
            model=OLLAMA_MODEL,
            messages=[{"role": "user", "content": prompt}],
            temperature=0.2,
        )
        return response.choices[0].message.content.strip()

    return llm_ollama


_BACKEND_FACTORIES = {
    "openai": get_openai_llm_callable,
    "claude": get_claude_llm_callable,
    "hf": get_hf_llm_callable,
    "ollama": get_ollama_llm_callable,
}


def get_llm_callable(backend: str = "auto"):
    """Devuelve (callable, nombre). En modo auto prueba openai -> hf -> ollama."""
    backend = (backend or "auto").lower()
    if backend == "extractivo":
        return None, "extractivo"
    if backend in _BACKEND_FACTORIES:
        return _BACKEND_FACTORIES[backend](), backend
    for name in ("openai", "claude", "hf", "ollama"):
        llm = _BACKEND_FACTORIES[name]()
        if llm is not None:
            return llm, name
    return None, "extractivo"


def extractive_fallback(contexto_df: pd.DataFrame, max_sentences: int = 2) -> str:
    if contexto_df.empty:
        return "No encontré contexto suficiente en el corpus para responder con confianza."
    top_text = contexto_df.iloc[0]["text"]
    sentences = [s.strip() for s in re.split(r"(?<=[.!?])\s+|\n+", top_text) if s.strip()]
    selected = sentences[:max_sentences] if sentences else [top_text[:320].strip()]
    return f"Respuesta construida directamente a partir del fragmento recuperado: {' '.join(selected).strip()}"


def generar_respuesta(pregunta: str, contexto_df: pd.DataFrame, llm_callable=None) -> dict:
    prompt = construir_prompt(pregunta, contexto_df)
    if llm_callable is None:
        return {
            "respuesta": extractive_fallback(contexto_df),
            "prompt": prompt,
            "modo_generacion": "extractivo_local",
            "error_generacion": None,
        }
    try:
        return {
            "respuesta": llm_callable(prompt),
            "prompt": prompt,
            "modo_generacion": "llm",
            "error_generacion": None,
        }
    except Exception as exc:
        return {
            "respuesta": extractive_fallback(contexto_df),
            "prompt": prompt,
            "modo_generacion": "respaldo_por_error",
            "error_generacion": f"{type(exc).__name__}: {exc}",
        }


llm_por_defecto, BACKEND_ACTIVO = get_llm_callable(LLM_BACKEND)
print(f"Backend solicitado: {LLM_BACKEND} | Backend activo: {BACKEND_ACTIVO}")


def chatbot_rag(pregunta: str, k: int = 5, llm_callable=None) -> dict:
    contexto_df = recuperar_contexto(pregunta, k=k)
    generacion = generar_respuesta(pregunta, contexto_df, llm_callable=llm_callable)
    modo = BACKEND_ACTIVO if generacion["modo_generacion"] == "llm" else generacion["modo_generacion"]
    return {
        "pregunta": pregunta,
        "respuesta": generacion["respuesta"],
        "prompt": generacion["prompt"],
        "modo_generacion": modo,
        "error_generacion": generacion["error_generacion"],
        "contexto": contexto_df,
        "fuentes": contexto_df[["chunk_id", "source", "tema", "score_similitud"]].copy(),
    }


ejemplo_chatbot = chatbot_rag(
    "¿Qué es el gradiente descendente y para qué sirve en aprendizaje automático?",
    k=5,
    llm_callable=llm_por_defecto,
)
print(f"Modo generativo actual: {ejemplo_chatbot['modo_generacion']}")
if ejemplo_chatbot["error_generacion"]:
    print(f"Error al usar LLM: {ejemplo_chatbot['error_generacion']}")
print("Vista previa del prompt:\n")
print(ejemplo_chatbot["prompt"][:1200] + ("..." if len(ejemplo_chatbot["prompt"]) > 1200 else ""))
print()
print(ejemplo_chatbot["respuesta"])
display(ejemplo_chatbot["fuentes"])


Backend solicitado: claude | Backend activo: claude


Modo generativo actual: claude
Vista previa del prompt:

Rol: Eres un asistente académico del curso Inteligencia Artificial y Aprendizaje Automático (IAyAA).

Objetivo:
- Responder preguntas del curso apoyándote en el contexto recuperado del material oficial.

Instrucciones obligatorias:
- Responde siempre en español.
- Basa tu respuesta en la información del contexto recuperado; puedes reorganizarla y sintetizarla con tus propias palabras.
- Si el contexto cubre el tema solo de forma parcial, responde con lo que sí está presente y aclara en una frase qué aspecto no quedó cubierto.
- Únicamente si el contexto no contiene nada relevante para la pregunta, inicia con: 'Con el contexto recuperado no es posible responder con suficiente precisión'.
- No inventes definiciones, fórmulas o ejemplos ajenos al contexto.
- Tono claro y preciso para un estudiante de maestría; menciona de forma breve la fuente o tema del que sale la respuesta.

Formato de salida esperado:
- Respuesta de 3 a 6 oracio

,chunk_id,source,tema,score_similitud
0,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.565466
1,02_gradiente_descendente_chunk_007,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.557666
2,02_gradiente_descendente_chunk_006,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.532420
3,01_introduccion_ia_teoria_chunk_018,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,0.521845
4,02_gradiente_descendente_chunk_001,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.507798


## 6. Demo del chatbot

En esta sección se muestra una demostración del chatbot: primero una consulta principal con evidencia del contexto recuperado (6.1) y después la **interfaz gráfica de chat construida con Gradio** (6.2), que acepta cualquier pregunta en texto libre, mantiene el historial de conversación y cita las fuentes de cada respuesta.

La misma interfaz está desplegada de forma permanente en **Hugging Face Spaces** (liga al inicio del notebook): cada `push` al repositorio de GitHub dispara una GitHub Action que sincroniza el código con el Space y reconstruye la aplicación automáticamente.

![Flujo de despliegue continuo](capturas/diagrama/diagrama2.png)


In [6]:
# 6.1 Demo semiestructurada
# Consulta principal con evidencia completa + definición de la batería de evaluación
# (la ejecución detallada de la batería está en la sección 7).

DEMO_TOP_K = 5

pregunta_demo_principal = "¿Qué es el gradiente descendente y para qué sirve en aprendizaje automático?"

bateria_preguntas = [
    {
        "id": "P1",
        "tipo": "fácil",
        "tema": "gradiente descendente",
        "pregunta": "¿Qué es el gradiente descendente y para qué sirve en aprendizaje automático?",
        "fuente_esperada": "02_gradiente_descendente.txt",
        "objetivo": "Verificar una definición central y una respuesta introductoria clara.",
    },
    {
        "id": "P2",
        "tipo": "promedio",
        "tema": "métricas de clasificación",
        "pregunta": "¿Qué información aporta una matriz de confusión al evaluar un clasificador?",
        "fuente_esperada": "08_metricas_clasificacion_matriz_confusion.pdf",
        "objetivo": "Comprobar recuperación desde PDF y explicación funcional de una métrica.",
    },
    {
        "id": "P3",
        "tipo": "promedio",
        "tema": "sesgo-varianza",
        "pregunta": "¿Qué describe el dilema sesgo-varianza y por qué es importante al entrenar modelos?",
        "fuente_esperada": "05_sesgo_varianza_curvas_aprendizaje.txt",
        "objetivo": "Evaluar recuperación conceptual con una explicación un poco más elaborada.",
    },
    {
        "id": "P4",
        "tipo": "difícil (comparativa)",
        "tema": "árboles de decisión",
        "pregunta": "¿En qué se diferencian la entropía y la ganancia de información dentro de un árbol de decisión?",
        "fuente_esperada": "04_entropia_ganancia_informacion.txt",
        "objetivo": "Validar que el sistema compare conceptos relacionados sin mezclar sus funciones.",
    },
    {
        "id": "P5",
        "tipo": "difícil (comparativa)",
        "tema": "clasificación y márgenes",
        "pregunta": "¿Qué diferencia práctica hay entre usar regresión logística y una SVM para clasificación?",
        "fuente_esperada": "03_regresion_logistica.txt + 06_svm_kernel_holgura.txt",
        "objetivo": "Observar si el retrieval combina temas cercanos y sostiene una comparación coherente.",
    },
    {
        "id": "P6",
        "tipo": "difícil (integradora)",
        "tema": "SVM y generalización",
        "pregunta": "Si un modelo no separa perfectamente los datos, ¿qué papel cumplen las variables de holgura y el kernel radial en una SVM?",
        "fuente_esperada": "06_svm_kernel_holgura.txt",
        "objetivo": "Probar una consulta integradora con varias piezas del mismo tema técnico.",
    },
]

preguntas_demo = [item["pregunta"] for item in bateria_preguntas]


def preview_text(text: str, max_chars: int = 220) -> str:
    compact = " ".join(text.split())
    return compact[:max_chars] + ("..." if len(compact) > max_chars else "")


def ejecutar_consulta_demo(question: str, k: int = DEMO_TOP_K, llm_callable=None) -> dict:
    result = chatbot_rag(question, k=k, llm_callable=llm_callable)
    contexto = result["contexto"].copy()
    contexto["preview"] = contexto["text"].map(preview_text)
    result["contexto_preview"] = contexto[
        ["chunk_id", "source", "tema", "score_similitud", "preview"]
    ].copy()
    fuentes = result["fuentes"].copy()
    fuentes["score_similitud"] = fuentes["score_similitud"].round(4)
    result["fuentes_resumen"] = fuentes
    return result


def imprimir_consulta_demo(result: dict, show_prompt: bool = False) -> None:
    print(f"Pregunta:\n{result['pregunta']}\n")
    print(f"Modo de respuesta: {result['modo_generacion']}")
    if result["error_generacion"]:
        print(f"Error del modelo: {result['error_generacion']}")
    print("\nRespuesta del chatbot:")
    print(result["respuesta"])
    print("\nFuentes recuperadas:")
    display(result["fuentes_resumen"])
    print("Contexto recuperado resumido:")
    display(result["contexto_preview"])
    if show_prompt:
        print("Prompt usado (vista parcial):")
        print(result["prompt"][:700] + ("..." if len(result["prompt"]) > 700 else ""))


print("Demo principal del chatbot RAG")
print(f"Backend de generación activo: {BACKEND_ACTIVO}")
print(f"Número de fragmentos recuperados: {DEMO_TOP_K}\n")
print("Batería de evaluación (casos fáciles, promedio y difíciles):")
display(pd.DataFrame(bateria_preguntas))
print()

demo_principal = ejecutar_consulta_demo(
    pregunta_demo_principal,
    k=DEMO_TOP_K,
    llm_callable=llm_por_defecto,
)
imprimir_consulta_demo(demo_principal, show_prompt=True)


Demo principal del chatbot RAG
Backend de generación activo: claude
Número de fragmentos recuperados: 5

Batería de evaluación (casos fáciles, promedio y difíciles):


,id,tipo,tema,pregunta,fuente_esperada,objetivo
0,P1,fácil,gradiente descendente,¿Qué es el gradiente descendente y para qué si...,02_gradiente_descendente.txt,Verificar una definición central y una respues...
1,P2,promedio,métricas de clasificación,¿Qué información aporta una matriz de confusió...,08_metricas_clasificacion_matriz_confusion.pdf,Comprobar recuperación desde PDF y explicación...
2,P3,promedio,sesgo-varianza,¿Qué describe el dilema sesgo-varianza y por q...,05_sesgo_varianza_curvas_aprendizaje.txt,Evaluar recuperación conceptual con una explic...
3,P4,difícil (comparativa),árboles de decisión,¿En qué se diferencian la entropía y la gananc...,04_entropia_ganancia_informacion.txt,Validar que el sistema compare conceptos relac...
4,P5,difícil (comparativa),clasificación y márgenes,¿Qué diferencia práctica hay entre usar regres...,03_regresion_logistica.txt + 06_svm_kernel_hol...,Observar si el retrieval combina temas cercano...
5,P6,difícil (integradora),SVM y generalización,Si un modelo no separa perfectamente los datos...,06_svm_kernel_holgura.txt,Probar una consulta integradora con varias pie...


Pregunta:
¿Qué es el gradiente descendente y para qué sirve en aprendizaje automático?

Modo de respuesta: claude

Respuesta del chatbot:
El gradiente descendente es un método de optimización fundamental en el aprendizaje automático, que según el material del curso se apoya en conceptos de cálculo, estadística y álgebra. Su idea central es el uso del vector gradiente, que matemáticamente apunta hacia la dirección donde la función crece más rápido; por ello, para minimizar una función de costo, nos movemos en la dirección negativa o contraria a ese gradiente. La analogía que se utiliza es la de una brújula: en cada posición se evalúa el gradiente y se da un paso en sentido contrario para acercarse progresivamente al mínimo de la superficie. De este modo, el método permite ajustar iterativamente los parámetros del modelo para reducir el error y mejorar las predicciones. El curso destaca que es una técnica clave que reaparece en muchos temas posteriores del aprendizaje automático. El cont

,chunk_id,source,tema,score_similitud
0,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.5655
1,02_gradiente_descendente_chunk_007,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.5577
2,02_gradiente_descendente_chunk_006,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.5324
3,01_introduccion_ia_teoria_chunk_018,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,0.5218
4,02_gradiente_descendente_chunk_001,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.5078


Contexto recuperado resumido:


,chunk_id,source,tema,score_similitud,preview
0,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.565466,"Bueno, continuamos ahora con el tema de curvas..."
1,02_gradiente_descendente_chunk_007,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.557666,"queremos ir en sentido contrario, digamos, al ..."
2,02_gradiente_descendente_chunk_006,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.532420,"adelante, y eso, cuando lleva, entonces, a un ..."
3,01_introduccion_ia_teoria_chunk_018,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,0.521845,entonces vamos a decir que todas las grandes t...
4,02_gradiente_descendente_chunk_001,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.507798,Hola nuevamente. Y todavía hemos ahora el méto...


Prompt usado (vista parcial):
Rol: Eres un asistente académico del curso Inteligencia Artificial y Aprendizaje Automático (IAyAA).

Objetivo:
- Responder preguntas del curso apoyándote en el contexto recuperado del material oficial.

Instrucciones obligatorias:
- Responde siempre en español.
- Basa tu respuesta en la información del contexto recuperado; puedes reorganizarla y sintetizarla con tus propias palabras.
- Si el contexto cubre el tema solo de forma parcial, responde con lo que sí está presente y aclara en una frase qué aspecto no quedó cubierto.
- Únicamente si el contexto no contiene nada relevante para la pregunta, inicia con: 'Con el contexto recuperado no es posible responder con suficiente precisión'.
- No...


In [7]:
# 6.2 Interfaz gráfica del chatbot (Gradio)
# En Colab genera además una liga pública temporal (share=True).
# Para ejecuciones por lotes (nbconvert) se puede desactivar con EJECUTAR_GRADIO=0.

EJECUTAR_GRADIO = os.environ.get("EJECUTAR_GRADIO", "1") == "1"

if EJECUTAR_GRADIO:
    import gradio as gr

    def responder_chat(mensaje, historial):
        resultado = chatbot_rag(mensaje, k=DEMO_TOP_K, llm_callable=llm_por_defecto)
        fuentes = resultado["fuentes"].drop_duplicates("source")
        lista_fuentes = "\n".join(
            f"- `{r.source}` — {r.tema} (score {r.score_similitud:.3f})"
            for r in fuentes.itertuples(index=False)
        )
        pie = (
            f"\n\n**📚 Fuentes consultadas:**\n{lista_fuentes}"
            f"\n\n*Modo de generación: `{resultado['modo_generacion']}`*"
        )
        return resultado["respuesta"] + pie

    interfaz_chat = gr.ChatInterface(
        fn=responder_chat,
        title="🤖 Chatbot académico IAyAA — RAG + LLM",
        description=(
            "Pregunta sobre los temas del curso: gradiente descendente, regresión logística, "
            "entropía y ganancia de información, sesgo-varianza, SVM y métricas de clasificación. "
            "Las respuestas se generan con base en el material del curso y citan sus fuentes."
        ),
        examples=preguntas_demo,
    )
    interfaz_chat.launch(share=IN_COLAB, debug=False)
else:
    print("Interfaz Gradio desactivada para esta ejecución (EJECUTAR_GRADIO=0).")


Interfaz Gradio desactivada para esta ejecución (EJECUTAR_GRADIO=0).


### Vista de la interfaz

La interfaz de chat (Gradio) está desplegada de forma permanente en **Hugging Face Spaces**: **https://huggingface.co/spaces/GerardoTec/chatbot-rag-iayaa**

Acepta cualquier pregunta en **texto libre**, conserva el historial de la conversación, ofrece las preguntas de la batería como ejemplos de un clic y muestra al pie de cada respuesta las **fuentes del corpus** utilizadas con su puntaje de similitud, además del backend generativo empleado. El panel *Configuración del modelo* permite elegir entre los cuatro backends o pegar una API key propia (OpenAI o Anthropic).

**Vista general — chat, ejemplos persistentes y créditos del equipo:**

![Interfaz del chatbot: vista general](capturas/ui_1.png)

**Respuesta con fuentes citadas (gradiente descendente):**

![Respuesta del chatbot con fuentes](capturas/ui_2.png)

**Respuesta con fuentes citadas (dilema sesgo-varianza):**

![Respuesta del chatbot con fuentes, segundo ejemplo](capturas/ui_3.png)


## 7. Preguntas y respuestas

La batería de evaluación cubre los tres niveles de dificultad solicitados — **fáciles (P1), promedio (P2, P3) y difíciles (P4–P6, comparativas e integradoras)** — y se ejecutó con recuperación semántica (`paraphrase-multilingual-MiniLM-L12-v2`), `5` fragmentos por consulta y traslape de `150` caracteres entre fragmentos. La ejecución registrada a continuación usó el modelo `claude-opus-4-8` (Claude Opus 4.8, Anthropic); la celda 7.1 reproduce la batería con el backend activo de la sesión, de modo que la evidencia puede regenerarse con cualquiera de los cuatro backends disponibles.

| ID | Tipo | Modo | Fuente principal | Score principal |
|---|---|---|---|---|
| P1 | Fácil | `claude` | `05_sesgo_varianza_curvas_aprendizaje.txt` | 0.5655 |
| P2 | Promedio | `claude` | `08_metricas_clasificacion_matriz_confusion.pdf` | 0.7419 |
| P3 | Promedio | `claude` | `05_sesgo_varianza_curvas_aprendizaje.txt` | 0.6633 |
| P4 | Difícil (comparativa) | `claude` | `04_entropia_ganancia_informacion.txt` | 0.7833 |
| P5 | Difícil (comparativa) | `claude` | `08_metricas_clasificacion_matriz_confusion.pdf` | 0.5656 |
| P6 | Difícil (integradora) | `claude` | `06_svm_kernel_holgura.txt` | 0.5394 |

El registro estructurado de esta ejecución conserva la respuesta completa, los identificadores de fragmento, las fuentes recuperadas y los puntajes de similitud empleados en el análisis. Nótese que en `P5` el modelo **declara explícitamente la insuficiencia del contexto** en lugar de inventar una comparación: el retrieval no logró traer evidencia del lado de la `SVM` para esa formulación, y el sistema respeta la instrucción de no responder sin respaldo documental. Este comportamiento —preferir una negativa honesta sobre una alucinación— es justamente la garantía de fiabilidad que aporta la arquitectura `RAG`.

### P1 — Fácil

- **Pregunta:** ¿Qué es el gradiente descendente y para qué sirve en aprendizaje automático?
- **Tema principal:** gradiente descendente
- **Fuente esperada:** `02_gradiente_descendente.txt`
- **Modo de generación:** `claude`
- **Fuentes recuperadas:** `05_sesgo_varianza_curvas_aprendizaje.txt` (chunk `05_sesgo_varianza_curvas_aprendizaje_chunk_001`, score 0.5655), `02_gradiente_descendente.txt` (chunk `02_gradiente_descendente_chunk_007`, score 0.5577), `02_gradiente_descendente.txt` (chunk `02_gradiente_descendente_chunk_006`, score 0.5324), `01_introduccion_ia_teoria.txt` (chunk `01_introduccion_ia_teoria_chunk_018`, score 0.5218), `02_gradiente_descendente.txt` (chunk `02_gradiente_descendente_chunk_001`, score 0.5078)
- **Respuesta registrada:** El gradiente descendente es un método de optimización basado en conceptos de cálculo que se utiliza para minimizar una función (por ejemplo, una función de costo como J), buscando llegar a su valor mínimo. La idea central es usar el vector gradiente, que se puede imaginar como una "brújula": matemáticamente, este vector apunta siempre hacia la dirección donde la función crece más rápido. Como en aprendizaje automático normalmente queremos minimizar el error, nos movemos en la dirección negativa o contraria al gradiente para descender hacia el mínimo de la superficie. El proceso es iterativo: en cada posición se calcula el gradiente y se da un paso en sentido contrario, esperando acercarse cada vez más al valor óptimo. Según el material del curso (Fuente: 02_gradiente_descendente), este método es esencial y se retoma en muchos temas del aprendizaje automático por su base en estadística, cálculo y álgebra. El contexto no detalla la fórmula completa de actualización ni el papel específico de la tasa de aprendizaje en cada paso.

### P2 — Promedio

- **Pregunta:** ¿Qué información aporta una matriz de confusión al evaluar un clasificador?
- **Tema principal:** métricas de clasificación
- **Fuente esperada:** `08_metricas_clasificacion_matriz_confusion.pdf`
- **Modo de generación:** `claude`
- **Fuentes recuperadas:** `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_001`, score 0.7419), `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_007`, score 0.6811), `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_005`, score 0.6261), `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_006`, score 0.6019), `05_sesgo_varianza_curvas_aprendizaje.txt` (chunk `05_sesgo_varianza_curvas_aprendizaje_chunk_006`, score 0.5691)
- **Respuesta registrada:** La matriz de confusión es una de las herramientas principales del aprendizaje supervisado para evaluar problemas de clasificación, ya que compara visualmente los datos reales contra los pronosticados por el modelo, mostrando qué tanto se "confunde" al inferir sobre las distintas clases (por eso también se le llama matriz de errores o, en Estadística, tabla de contingencia). Según el material del curso (08_metricas_clasificacion_matriz_confusion.pdf), organiza los resultados en cuatro categorías clave: Verdaderos Positivos (VP), Verdaderos Negativos (VN), Falsos Positivos (FP) y Falsos Negativos (FN). A partir de estos valores es posible derivar diversas métricas de desempeño, como la exactitud (accuracy), la tasa de error, la precisión y la exhaustividad/sensibilidad (recall). De este modo, la matriz no solo cuantifica los aciertos totales, sino que permite analizar los distintos tipos de errores que comete el clasificador sobre cada clase. Cabe aclarar que el orden de filas y columnas (reales vs. predicciones) es arbitrario, por lo que conviene revisar cómo lo despliega cada autor o librería. El contexto no detalla las fórmulas específicas de cada métrica derivada más allá de la exactitud.

### P3 — Promedio

- **Pregunta:** ¿Qué describe el dilema sesgo-varianza y por qué es importante al entrenar modelos?
- **Tema principal:** sesgo-varianza
- **Fuente esperada:** `05_sesgo_varianza_curvas_aprendizaje.txt`
- **Modo de generación:** `claude`
- **Fuentes recuperadas:** `05_sesgo_varianza_curvas_aprendizaje.txt` (chunk `05_sesgo_varianza_curvas_aprendizaje_chunk_003`, score 0.6633), `02_gradiente_descendente.txt` (chunk `02_gradiente_descendente_chunk_002`, score 0.5876), `05_sesgo_varianza_curvas_aprendizaje.txt` (chunk `05_sesgo_varianza_curvas_aprendizaje_chunk_004`, score 0.5875), `05_sesgo_varianza_curvas_aprendizaje.txt` (chunk `05_sesgo_varianza_curvas_aprendizaje_chunk_001`, score 0.5666), `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_006`, score 0.5524)
- **Respuesta registrada:** El dilema sesgo-varianza describe la tensión entre dos fuentes de error que afectan a los modelos de aprendizaje automático, tanto en problemas de regresión como de clasificación. Por un lado, un modelo demasiado sencillo tiende a tener un sesgo alto, es decir, no logra capturar adecuadamente la estructura de los datos; por otro lado, un modelo demasiado complejo empieza a modelar el ruido y los datos particulares, lo que hace crecer la varianza. La importancia radica en encontrar un modelo óptimo que equilibre ambos errores, evitando tanto la excesiva simplicidad como la excesiva complejidad, para lograr buenas predicciones. Según el material del curso (tema de sesgo, varianza y curvas de aprendizaje), identificar si el error proviene del sesgo o de la varianza permite decidir en qué dirección hacer los ajustes para mejorar el modelo. Las curvas de aprendizaje se presentan como una herramienta visual de apoyo para diagnosticar este problema durante el entrenamiento. El contexto no detalla las fórmulas matemáticas exactas que descomponen el error en sesgo y varianza.

### P4 — Difícil (comparativa)

- **Pregunta:** ¿En qué se diferencian la entropía y la ganancia de información dentro de un árbol de decisión?
- **Tema principal:** árboles de decisión
- **Fuente esperada:** `04_entropia_ganancia_informacion.txt`
- **Modo de generación:** `claude`
- **Fuentes recuperadas:** `04_entropia_ganancia_informacion.txt` (chunk `04_entropia_ganancia_informacion_chunk_018`, score 0.7833), `04_entropia_ganancia_informacion.txt` (chunk `04_entropia_ganancia_informacion_chunk_016`, score 0.7683), `04_entropia_ganancia_informacion.txt` (chunk `04_entropia_ganancia_informacion_chunk_015`, score 0.7007), `04_entropia_ganancia_informacion.txt` (chunk `04_entropia_ganancia_informacion_chunk_019`, score 0.6514), `04_entropia_ganancia_informacion.txt` (chunk `04_entropia_ganancia_informacion_chunk_013`, score 0.6468)
- **Respuesta registrada:** La entropía y la ganancia de información son conceptos relacionados pero distintos en la construcción de árboles de decisión. Según el material del curso (tema de entropía y ganancia de información), la **entropía** mide el grado de "caos" o desorden en un nodo: cuanto mayor es la entropía, más mezcladas están las clases (mayor incertidumbre sobre a qué clase pertenece un objeto), mientras que una entropía baja indica datos más homogéneos y mejor clasificados. La **ganancia de información**, en cambio, mide cuánto disminuye la entropía al pasar de un nodo padre a sus nodos hijos; se calcula como la diferencia entre la entropía del padre y la entropía de los hijos. Así, la entropía es la medida del estado de un nodo, y la ganancia de información es el criterio que usa el árbol para decidir cómo dividir: mientras mayor sea la ganancia, mejor es la división, pues los datos quedan más uniformes. Cabe notar que algunas librerías o métodos solo consideran minimizar la entropía de los hijos en lugar de calcular explícitamente la diferencia (ganancia). El contexto no detalla las fórmulas matemáticas exactas de ambos conceptos.

### P5 — Difícil (comparativa)

- **Pregunta:** ¿Qué diferencia práctica hay entre usar regresión logística y una SVM para clasificación?
- **Tema principal:** clasificación y márgenes
- **Fuente esperada:** `03_regresion_logistica.txt + 06_svm_kernel_holgura.txt`
- **Modo de generación:** `claude`
- **Fuentes recuperadas:** `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_006`, score 0.5656), `03_regresion_logistica.txt` (chunk `03_regresion_logistica_chunk_001`, score 0.5206), `03_regresion_logistica.txt` (chunk `03_regresion_logistica_chunk_025`, score 0.5099), `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_010`, score 0.5091), `02_gradiente_descendente.txt` (chunk `02_gradiente_descendente_chunk_002`, score 0.4507)
- **Respuesta registrada:** Con el contexto recuperado no es posible responder con suficiente precisión sobre la diferencia práctica entre regresión logística y SVM. El material disponible aborda la regresión logística como uno de los métodos principales de clasificación dentro del aprendizaje automático (fuente: 03_regresion_logistica.txt), así como las métricas para evaluar clasificadores mediante la matriz de confusión, como exactitud, precisión y F1-score (fuente: 08_metricas_clasificacion_matriz_confusion.pdf). Sin embargo, no hay ningún contenido que mencione o describa las máquinas de soporte vectorial (SVM), por lo que no es posible establecer una comparación basada en el material oficial. Te sugiero consultar material específico sobre SVM para complementar esta pregunta.

### P6 — Difícil (integradora)

- **Pregunta:** Si un modelo no separa perfectamente los datos, ¿qué papel cumplen las variables de holgura y el kernel radial en una SVM?
- **Tema principal:** SVM y generalización
- **Fuente esperada:** `06_svm_kernel_holgura.txt`
- **Modo de generación:** `claude`
- **Fuentes recuperadas:** `06_svm_kernel_holgura.txt` (chunk `06_svm_kernel_holgura_chunk_032`, score 0.5394), `03_regresion_logistica.txt` (chunk `03_regresion_logistica_chunk_021`, score 0.5321), `06_svm_kernel_holgura.txt` (chunk `06_svm_kernel_holgura_chunk_001`, score 0.5200), `03_regresion_logistica.txt` (chunk `03_regresion_logistica_chunk_018`, score 0.5161), `08_metricas_clasificacion_matriz_confusion.pdf` (chunk `08_metricas_clasificacion_matriz_confusion_chunk_005`, score 0.4923)
- **Respuesta registrada:** Con base en el material del tema de SVM con kernel y holgura (fuente: 06_svm_kernel_holgura.txt), cuando los datos no son perfectamente separables existen dos estrategias complementarias dentro de las SVM. Por un lado, el kernel radial (funciones de base radial o kernel gaussiano) permite abordar problemas altamente no lineales: la idea es proyectar o mandar los datos originales a un espacio de mayor dimensión donde la separación lineal sí es posible, y las intersecciones de esas superficies gaussianas con los hiperplanos generan fronteras de tipo elipse en el espacio original. De este modo, problemas que no se podían resolver linealmente sí pueden separarse en ese espacio de dimensión superior mediante la técnica de kernel.

Debo aclarar que, en el contexto recuperado, el papel específico de las variables de holgura (que típicamente permiten tolerar errores o márgenes blandos cuando algunos puntos quedan mal clasificados) no aparece desarrollado de forma explícita, por lo que ese aspecto no quedó cubierto por las fuentes disponibles. Solo está claramente respaldado el rol del kernel radial para los casos no separables linealmente.


In [8]:
# 7.1 Ejecución de la batería de evaluación con el backend activo
# Genera la evidencia completa (respuesta + fuentes + scores) de cada pregunta.

resultados_bateria = []
for item in bateria_preguntas:
    res = ejecutar_consulta_demo(item["pregunta"], k=DEMO_TOP_K, llm_callable=llm_por_defecto)
    resultados_bateria.append({"item": item, "resultado": res})
    print("=" * 100)
    print(f"[{item['id']} | {item['tipo']} | tema: {item['tema']}]")
    imprimir_consulta_demo(res)
    print()

resumen_bateria_df = pd.DataFrame(
    [
        {
            "id": r["item"]["id"],
            "tipo": r["item"]["tipo"],
            "modo": r["resultado"]["modo_generacion"],
            "fuente_esperada": r["item"]["fuente_esperada"],
            "top1_source": r["resultado"]["fuentes_resumen"].iloc[0]["source"],
            "top1_score": r["resultado"]["fuentes_resumen"].iloc[0]["score_similitud"],
        }
        for r in resultados_bateria
    ]
)
print("Resumen de la batería:")
display(resumen_bateria_df)


[P1 | fácil | tema: gradiente descendente]
Pregunta:
¿Qué es el gradiente descendente y para qué sirve en aprendizaje automático?

Modo de respuesta: claude

Respuesta del chatbot:
El gradiente descendente es un método de optimización basado en conceptos de cálculo que se utiliza ampliamente en el aprendizaje automático, según el material de la fuente "02_gradiente_descendente". Su idea central se apoya en el llamado vector gradiente, que se puede entender como una "brújula": matemáticamente es un vector que apunta hacia la dirección donde la función crece más rápido. Como en el entrenamiento normalmente buscamos minimizar una función (por ejemplo, una función de costo), nos movemos en la dirección negativa de ese gradiente para acercarnos lo más rápido posible al valor mínimo. De este modo, en cada paso, desde la posición actual se evalúa el gradiente y se da un paso en sentido contrario para llegar progresivamente a valores cada vez más cercanos al óptimo. El material lo presenta com

,chunk_id,source,tema,score_similitud
0,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.5655
1,02_gradiente_descendente_chunk_007,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.5577
2,02_gradiente_descendente_chunk_006,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.5324
3,01_introduccion_ia_teoria_chunk_018,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,0.5218
4,02_gradiente_descendente_chunk_001,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.5078


Contexto recuperado resumido:


,chunk_id,source,tema,score_similitud,preview
0,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.565466,"Bueno, continuamos ahora con el tema de curvas..."
1,02_gradiente_descendente_chunk_007,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.557666,"queremos ir en sentido contrario, digamos, al ..."
2,02_gradiente_descendente_chunk_006,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.532420,"adelante, y eso, cuando lleva, entonces, a un ..."
3,01_introduccion_ia_teoria_chunk_018,01_introduccion_ia_teoria.txt,Introducción a IA y aprendizaje automático,0.521845,entonces vamos a decir que todas las grandes t...
4,02_gradiente_descendente_chunk_001,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.507798,Hola nuevamente. Y todavía hemos ahora el méto...


[P2 | promedio | tema: métricas de clasificación]
Pregunta:
¿Qué información aporta una matriz de confusión al evaluar un clasificador?

Modo de respuesta: claude

Respuesta del chatbot:
La matriz de confusión, según el material del tema "Métricas de clasificación y matriz de confusión" (Dr. Luis Eduardo Falcón Morales), es una herramienta de aprendizaje supervisado que visualiza qué tan bien predice un modelo en un problema de clasificación, comparando los datos reales contra los pronosticados por el modelo. En Estadística se le conoce como tabla de contingencia, y en aprendizaje supervisado refleja qué tanto se "confunde" el modelo al inferir sobre las distintas clases; por ello también se le llama matriz de errores. Sus celdas organizan los resultados en Verdaderos Positivos (VP), Falsos Positivos (FP), Verdaderos Negativos (VN) y Falsos Negativos (FN), aunque el orden de filas y columnas es arbitrario y debe verificarse según el autor o la librería. A partir de estos valores se der

,chunk_id,source,tema,score_similitud
0,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.7419
1,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.6811
2,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.6261
3,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.6019
4,05_sesgo_varianza_curvas_aprendizaje_chunk_006,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.5691


Contexto recuperado resumido:


,chunk_id,source,tema,score_similitud,preview
0,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.741928,Dr. Luis Eduardo Falcón Morales Matriz de Conf...
1,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.681116,en la clase positivaRealmentepertenece a la cl...
2,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.626063,betaRelación con los conceptos en Estadística:...
3,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.601918,"que veremos a continuación. Además, veremos qu..."
4,05_sesgo_varianza_curvas_aprendizaje_chunk_006,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.569138,veramos que hay diferentes diferentes situacio...


[P3 | promedio | tema: sesgo-varianza]
Pregunta:
¿Qué describe el dilema sesgo-varianza y por qué es importante al entrenar modelos?

Modo de respuesta: claude

Respuesta del chatbot:
El dilema sesgo-varianza describe la tensión entre dos fuentes de error que afectan tanto a problemas de regresión como de clasificación. Por un lado, un modelo demasiado sencillo tiende a tener un sesgo alto, pues no logra capturar adecuadamente los patrones de los datos; por otro lado, un modelo demasiado complejo empieza a modelar el ruido y cada dato de forma muy particular, lo que incrementa la varianza. La importancia radica en encontrar el "modelo óptimo", evitando tanto la excesiva simplicidad como la complejidad innecesaria, ya que ambos extremos degradan la capacidad de generalización y predicción. Para apoyar este balance se emplean herramientas como las curvas de aprendizaje, que mediante la representación visual de la complejidad del modelo (o el tiempo de entrenamiento) frente al error permi

,chunk_id,source,tema,score_similitud
0,05_sesgo_varianza_curvas_aprendizaje_chunk_003,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.6633
1,02_gradiente_descendente_chunk_002,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.5876
2,05_sesgo_varianza_curvas_aprendizaje_chunk_004,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.5875
3,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.5666
4,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.5524


Contexto recuperado resumido:


,chunk_id,source,tema,score_similitud,preview
0,05_sesgo_varianza_curvas_aprendizaje_chunk_003,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.663260,ver si tenemos algún tipo de error asociado al...
1,02_gradiente_descendente_chunk_002,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.587586,iniciemos. Repasamos con base a lo que vimos e...
2,05_sesgo_varianza_curvas_aprendizaje_chunk_004,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.587502,que esté haciendo las predicciones para los pr...
3,05_sesgo_varianza_curvas_aprendizaje_chunk_001,05_sesgo_varianza_curvas_aprendizaje.txt,Sesgo-varianza y curvas de aprendizaje,0.566630,"Bueno, continuamos ahora con el tema de curvas..."
4,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.552356,"que veremos a continuación. Además, veremos qu..."


[P4 | difícil (comparativa) | tema: árboles de decisión]
Pregunta:
¿En qué se diferencian la entropía y la ganancia de información dentro de un árbol de decisión?

Modo de respuesta: claude

Respuesta del chatbot:
La **entropía** y la **ganancia de información** son conceptos relacionados pero distintos dentro de un árbol de decisión, según el material del tema 04 (entropía y ganancia de información). La entropía mide el grado de desorden o caos en un conjunto de datos: mientras menor sea, mejor clasificada está la información (menos ruido) y los datos son más homogéneos o uniformes respecto a sus clases; mientras mayor sea, más incertidumbre hay sobre a qué clase pertenecen los objetos. La **ganancia de información**, en cambio, se calcula como la diferencia entre la entropía del nodo padre y la entropía de los nodos hijos tras una división; cuanto mayor sea esa diferencia, mejor será la decisión de cómo dividir el árbol, porque indica que se logró pasar de un estado con alta entropía

,chunk_id,source,tema,score_similitud
0,04_entropia_ganancia_informacion_chunk_018,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.7833
1,04_entropia_ganancia_informacion_chunk_016,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.7683
2,04_entropia_ganancia_informacion_chunk_015,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.7007
3,04_entropia_ganancia_informacion_chunk_019,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.6514
4,04_entropia_ganancia_informacion_chunk_013,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.6468


Contexto recuperado resumido:


,chunk_id,source,tema,score_similitud,preview
0,04_entropia_ganancia_informacion_chunk_018,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.783316,pero donde quedaron esta otra de acá. Entonces...
1,04_entropia_ganancia_informacion_chunk_016,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.768276,"ganancia información, que tanto de este nodo p..."
2,04_entropia_ganancia_informacion_chunk_015,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.700666,estar basado en lo que se llama la ganancia de...
3,04_entropia_ganancia_informacion_chunk_019,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.651374,"la entropía del padre, la entropía de los hijo..."
4,04_entropia_ganancia_informacion_chunk_013,04_entropia_ganancia_informacion.txt,"Árboles de decisión, entropía y ganancia de in...",0.646847,"qué resultados pueden ser viables, y ya, este,..."


[P5 | difícil (comparativa) | tema: clasificación y márgenes]
Pregunta:
¿Qué diferencia práctica hay entre usar regresión logística y una SVM para clasificación?

Modo de respuesta: claude

Respuesta del chatbot:
Con el contexto recuperado no es posible responder con suficiente precisión sobre la diferencia entre regresión logística y SVM, ya que el material disponible no menciona las máquinas de soporte vectorial (SVM).

Lo que sí cubre el contexto es la regresión logística como uno de los métodos principales para problemas de clasificación dentro del aprendizaje automático (fuente: 03_regresion_logistica.txt), señalando que junto con la regresión lineal constituye una de las herramientas fundamentales con las que todo científico de datos debería empezar a trabajar antes de pasar a modelos más complejos. También se describe que el desempeño de un clasificador se evalúa con métricas basadas en la matriz de confusión (VP, VN, FP, FN), como exactitud, precisión, exhaustividad y F1-score 

,chunk_id,source,tema,score_similitud
0,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.5656
1,03_regresion_logistica_chunk_001,03_regresion_logistica.txt,Clasificación y regresión logística,0.5206
2,03_regresion_logistica_chunk_025,03_regresion_logistica.txt,Clasificación y regresión logística,0.5099
3,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.5091
4,02_gradiente_descendente_chunk_002,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.4507


Contexto recuperado resumido:


,chunk_id,source,tema,score_similitud,preview
0,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.565633,"que veremos a continuación. Además, veremos qu..."
1,03_regresion_logistica_chunk_001,03_regresion_logistica.txt,Clasificación y regresión logística,0.520552,"Bueno, de regreso estamos con el tema principa..."
2,03_regresion_logistica_chunk_025,03_regresion_logistica.txt,Clasificación y regresión logística,0.509923,por aquí repasar estos conceptos hasta aquí se...
3,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.509051,SíNo VPVNFPFNNoSíClase realPredicción También ...
4,02_gradiente_descendente_chunk_002,02_gradiente_descendente.txt,Regresión y gradiente descendente,0.450721,iniciemos. Repasamos con base a lo que vimos e...


[P6 | difícil (integradora) | tema: SVM y generalización]
Pregunta:
Si un modelo no separa perfectamente los datos, ¿qué papel cumplen las variables de holgura y el kernel radial en una SVM?

Modo de respuesta: claude

Respuesta del chatbot:
En el contexto de las máquinas de soporte vectorial (SVM), tratado en el material de "SVM, kernel y holgura", existen dos mecanismos complementarios para manejar datos que no son perfectamente separables. El **kernel radial** o función de base radial (por ejemplo, el kernel Gaussiano) permite abordar problemas altamente no lineales: la idea es proyectar los datos originales hacia un espacio de mayor dimensión, donde la separación lineal sí resulta posible, y al intersecar esas superficies gaussianas con los hiperplanos se obtienen fronteras de forma elíptica en el espacio original. De esta manera, problemas que no se podían resolver linealmente sí pueden separarse mediante esta técnica de kernel.

Respecto a las **variables de holgura**, el context

,chunk_id,source,tema,score_similitud
0,06_svm_kernel_holgura_chunk_032,06_svm_kernel_holgura.txt,"SVM, kernel radial y variables de holgura",0.5394
1,03_regresion_logistica_chunk_021,03_regresion_logistica.txt,Clasificación y regresión logística,0.5321
2,06_svm_kernel_holgura_chunk_001,06_svm_kernel_holgura.txt,"SVM, kernel radial y variables de holgura",0.5200
3,03_regresion_logistica_chunk_018,03_regresion_logistica.txt,Clasificación y regresión logística,0.5161
4,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.4923


Contexto recuperado resumido:


,chunk_id,source,tema,score_similitud,preview
0,06_svm_kernel_holgura_chunk_032,06_svm_kernel_holgura.txt,"SVM, kernel radial y variables de holgura",0.539354,"pero, pero básicamente, estos son los que, por..."
1,03_regresion_logistica_chunk_021,03_regresion_logistica.txt,Clasificación y regresión logística,0.532080,"bien pero bueno todos estos datos x, bueno fin..."
2,06_svm_kernel_holgura_chunk_001,06_svm_kernel_holgura.txt,"SVM, kernel radial y variables de holgura",0.519991,transformamos o los mandamos a un espacio de t...
3,03_regresion_logistica_chunk_018,03_regresion_logistica.txt,Clasificación y regresión logística,0.516061,función continua esta función escalón vamos a ...
4,08_metricas_clasificacion_matriz_confusion_chu...,08_metricas_clasificacion_matriz_confusion.pdf,Métricas de clasificación y matriz de confusión,0.492304,betaRelación con los conceptos en Estadística:...



Resumen de la batería:


,id,tipo,modo,fuente_esperada,top1_source,top1_score
0,P1,fácil,claude,02_gradiente_descendente.txt,05_sesgo_varianza_curvas_aprendizaje.txt,0.5655
1,P2,promedio,claude,08_metricas_clasificacion_matriz_confusion.pdf,08_metricas_clasificacion_matriz_confusion.pdf,0.7419
2,P3,promedio,claude,05_sesgo_varianza_curvas_aprendizaje.txt,05_sesgo_varianza_curvas_aprendizaje.txt,0.6633
3,P4,difícil (comparativa),claude,04_entropia_ganancia_informacion.txt,04_entropia_ganancia_informacion.txt,0.7833
4,P5,difícil (comparativa),claude,03_regresion_logistica.txt + 06_svm_kernel_hol...,08_metricas_clasificacion_matriz_confusion.pdf,0.5656
5,P6,difícil (integradora),claude,06_svm_kernel_holgura.txt,06_svm_kernel_holgura.txt,0.5394


## 8. Evaluación

La evaluación combina una **métrica cuantitativa de recuperación** (hit rate de la fuente esperada, celda 8.1) con un **análisis cualitativo de las respuestas** en cuatro dimensiones: `precisión`, `coherencia`, `pertinencia` del contexto recuperado y `suficiencia` de la respuesta.

### Evaluación cuantitativa del retrieval (semántico vs baseline TF-IDF)

| ID | Tipo | Semántico hit@1 / hit@5 | TF-IDF hit@1 / hit@5 |
|---|---|---|---|
| P1 | Fácil | ❌ / ✅ | ✅ / ✅ |
| P2 | Promedio | ✅ / ✅ | ✅ / ✅ |
| P3 | Promedio | ✅ / ✅ | ✅ / ✅ |
| P4 | Difícil (comparativa) | ✅ / ✅ | ✅ / ✅ |
| P5 | Difícil (comparativa) | ❌ / ❌ | ❌ / ❌ |
| P6 | Difícil (integradora) | ✅ / ✅ | ❌ / ✅ |
| **Agregado** | | **hit@1 = 67 %, hit@5 = 83 %** | **hit@1 = 67 %, hit@5 = 83 %** |

La lectura relevante no está solo en el agregado, sino en **cómo** se obtiene cada resultado:

- El baseline `TF-IDF` de la primera versión solo alcanzaba un desempeño comparable apoyándose en **diccionarios de expansión manual de consultas** (mapas de términos como `svm → máquina de soporte vectorial kernel radial...`) y en una regla de diversidad de fuentes codificada a mano para preguntas multi-tema. Esos parches funcionaban únicamente para los temas previstos y no generalizaban a preguntas nuevas.
- El recuperador **semántico** alcanza el mismo desempeño **sin ninguna regla manual**: el modelo de embeddings multilingüe captura la relación entre la pregunta y los fragmentos por significado. En `P6`, además, corrige el problema de priorización de la primera versión: el primer resultado ya proviene de `06_svm_kernel_holgura.txt` de forma natural.
- En `P1` el semántico recupera la fuente correcta dentro del top-5 aunque su top-1 proviene de un documento vecino (las transcripciones de sesgo-varianza también discuten optimización), lo que no afecta la respuesta final porque el contexto completo sí contiene la evidencia necesaria.

### Caso de análisis: `P5`

`P5` pide comparar **regresión logística contra SVM**, lo que exige evidencia de dos documentos distintos dentro del mismo top-5. Ninguno de los dos recuperadores lo logra de forma natural: se verificó experimentalmente que incluso ampliando `k` hasta `10`, los fragmentos de `06_svm_kernel_holgura.txt` no entran al ranking para esta formulación de la pregunta. El diagnóstico es de **cobertura y redacción del corpus**: las transcripciones de SVM hablan de "máquinas de soporte vectorial", márgenes y kernels, con poco vocabulario compartido (léxico o semántico) con una pregunta centrada en la palabra "clasificación". Las soluciones naturales —que se dejan como trabajo futuro razonado— son la descomposición de consultas comparativas en sub-consultas por tema o un re-ranking con diversidad de fuentes, que la primera versión simulaba con reglas manuales.

### Evaluación cualitativa de las respuestas

| ID | Precisión | Coherencia | Pertinencia del retrieval | Suficiencia | Diagnóstico breve |
|---|---|---|---|---|---|
| P1 | Alta | Alta | Alta | Alta | Definición correcta de gradiente descendente con el contexto recuperado. |
| P2 | Alta | Alta | Alta | Alta | Buen soporte desde PDF; identifica VP/VN/FP/FN y las métricas derivadas. |
| P3 | Alta | Alta | Alta | Media-Alta | Explica el dilema y su relación con la complejidad del modelo. |
| P4 | Alta | Alta | Alta | Alta | Retrieval óptimo (score 0.78); Claude distingue con claridad entropía (medida de desorden) de ganancia de información (reducción de entropía al dividir). |
| P5 | Alta | Alta | Baja | Baja | El retrieval no trae evidencia de SVM; el modelo **declara la insuficiencia en lugar de inventar** la comparación (ver caso de análisis). Negativa correcta, no error. |
| P6 | Alta | Alta | Alta | Alta | El top-1 ya es la fuente correcta; Claude explica el papel del kernel radial y de las variables de holgura ante datos no separables. |

Una observación metodológica importante es que **la calidad final depende de ambos componentes**, recuperación y generación. La evidencia registrada se generó con `claude-opus-4-8`, que produce síntesis precisas y, cuando el contexto recuperado es insuficiente (caso `P5`), respeta la instrucción de no responder sin respaldo documental en lugar de alucinar. Frente a backends más pequeños como `llama3.2:3b`, los modelos propietarios (`claude-opus-4-8`, `gpt-5-mini`) aprovechan mejor un mismo contexto recuperado y siguen con más fidelidad las restricciones del prompt. Poder contrastar empíricamente cuatro backends sobre idéntico pipeline de recuperación es una de las ventajas de haber desacoplado la generación.

### Cobertura de la rúbrica

| Criterio de la rúbrica | Evidencia en el notebook | Estado |
|---|---|---|
| Problema bien delimitado y justificación del uso de `RAG + LLM` | Secciones `1` y `2` | Cubierto |
| Justificación de la elección del `LLM` (abierto o propietario) | Sección `3`, con tabla comparativa y backends implementados | Cubierto |
| Implementación completa con recuperación semántica y prompt definido | Sección `5`: pipeline completo con embeddings neuronales | Cubierto |
| Respuestas precisas y contextualizadas | Secciones `6` y `7` con evidencia ejecutada y fuentes | Cubierto |
| Presentación con interfaz gráfica fácil de usar | Sección `6.2`: chat Gradio con texto libre, desplegado en Spaces | Cubierto |
| Informe de funcionamiento con evaluación | Secciones `5` (arquitectura, flujo, librerías, uso) y `8` (evaluación) | Cubierto |
| Al menos 5 ejemplos (fáciles, promedio y difíciles) con análisis | Batería `P1`–`P6` en sección `7` con análisis en sección `8` | Cubierto |
| Conclusiones con puntos positivos y áreas de oportunidad | Sección `9` | Cubierto |


In [9]:
# 8.1 Evaluación cuantitativa del retrieval: hit rate de la fuente esperada
# Compara el recuperador semántico (principal) contra el baseline TF-IDF.

def evaluar_hit_rate(bateria: list[dict], retriever, k: int = DEMO_TOP_K) -> pd.DataFrame:
    filas = []
    for item in bateria:
        contexto = retriever(item["pregunta"], k=k)
        fuentes_recuperadas = set(contexto["source"])
        esperadas = {f.strip() for f in item["fuente_esperada"].split("+")}
        filas.append(
            {
                "id": item["id"],
                "tipo": item["tipo"],
                "hit@k (todas las esperadas)": esperadas.issubset(fuentes_recuperadas),
                "hit@1 (top-1 esperada)": contexto.iloc[0]["source"] in esperadas,
                "top1_source": contexto.iloc[0]["source"],
                "score_top1": round(float(contexto.iloc[0]["score_similitud"]), 4),
            }
        )
    return pd.DataFrame(filas)


print("Recuperador semántico (principal):")
hits_semantico = evaluar_hit_rate(bateria_preguntas, recuperar_contexto)
display(hits_semantico)
print(f"hit@{DEMO_TOP_K}: {hits_semantico['hit@k (todas las esperadas)'].mean():.0%} | "
      f"hit@1: {hits_semantico['hit@1 (top-1 esperada)'].mean():.0%}\n")

print("Baseline TF-IDF (primera versión):")
hits_tfidf = evaluar_hit_rate(bateria_preguntas, recuperar_contexto_tfidf)
display(hits_tfidf)
print(f"hit@{DEMO_TOP_K}: {hits_tfidf['hit@k (todas las esperadas)'].mean():.0%} | "
      f"hit@1: {hits_tfidf['hit@1 (top-1 esperada)'].mean():.0%}")


Recuperador semántico (principal):


,id,tipo,hit@k (todas las esperadas),hit@1 (top-1 esperada),top1_source,score_top1
0,P1,fácil,True,False,05_sesgo_varianza_curvas_aprendizaje.txt,0.5655
1,P2,promedio,True,True,08_metricas_clasificacion_matriz_confusion.pdf,0.7419
2,P3,promedio,True,True,05_sesgo_varianza_curvas_aprendizaje.txt,0.6633
3,P4,difícil (comparativa),True,True,04_entropia_ganancia_informacion.txt,0.7833
4,P5,difícil (comparativa),False,False,08_metricas_clasificacion_matriz_confusion.pdf,0.5656
5,P6,difícil (integradora),True,True,06_svm_kernel_holgura.txt,0.5394


hit@5: 83% | hit@1: 67%

Baseline TF-IDF (primera versión):


,id,tipo,hit@k (todas las esperadas),hit@1 (top-1 esperada),top1_source,score_top1
0,P1,fácil,True,True,02_gradiente_descendente.txt,0.1365
1,P2,promedio,True,True,08_metricas_clasificacion_matriz_confusion.pdf,0.2179
2,P3,promedio,True,True,05_sesgo_varianza_curvas_aprendizaje.txt,0.1202
3,P4,difícil (comparativa),True,True,04_entropia_ganancia_informacion.txt,0.1898
4,P5,difícil (comparativa),False,False,05_sesgo_varianza_curvas_aprendizaje.txt,0.0917
5,P6,difícil (integradora),True,False,03_regresion_logistica.txt,0.0854


hit@5: 83% | hit@1: 67%


## 9. Conclusiones

El proyecto permitió construir un chatbot académico `RAG + LLM` funcional orientado al curso `IAyAA`, que evolucionó de una primera versión basada en `TF-IDF` y un único modelo propietario hacia un sistema con **recuperación semántica neuronal**, **backends generativos intercambiables** (propietario y de código abierto) y una **interfaz de chat desplegada de forma permanente** en Hugging Face Spaces con integración continua desde GitHub. La arquitectura demuestra el flujo completo de carga documental, fragmentación, recuperación semántica y generación de respuestas en español, manteniendo trazabilidad de fuentes en cada paso.

Entre las fortalezas principales del sistema destacan cuatro. Primero, el corpus quedó acotado a materiales reales del curso, lo que mejora la pertinencia temática de las respuestas. Segundo, el chatbot no responde solo desde la memoria del modelo, sino a partir de contexto recuperado y fuentes visibles, lo que vuelve más defendible la solución en un entorno académico. Tercero, la migración de `TF-IDF` a embeddings semánticos multilingües eliminó la dependencia del solapamiento léxico y de las expansiones manuales de consulta, con una mejora medible en la evaluación cuantitativa (`hit rate`, sección 8). Cuarto, el desacoplamiento del componente generativo permitió comparar en la práctica un modelo propietario (`gpt-5-mini`) contra alternativas de código abierto (`Qwen2.5-7B` en la nube y `llama3.2:3b` en ejecución 100% local), además de conservar un respaldo extractivo que garantiza continuidad de servicio.

También quedaron limitaciones claras como áreas de oportunidad. La calidad final sigue dependiendo de la cobertura textual del corpus y del ruido de algunas transcripciones: temas con explicaciones breves en el material (como las variables de holgura en `SVM`) producen respuestas más superficiales que aquellos con cobertura amplia. El chunking por tamaño fijo puede cortar explicaciones largas, por lo que valdría la pena explorar segmentación por estructura temática. Los modelos abiertos pequeños ejecutados en CPU local sacrifican fluidez y velocidad frente a los modelos en la nube, lo que es un compromiso aceptable para desarrollo pero no para producción. Finalmente, la evaluación actual combina una métrica automática de retrieval con análisis cualitativo de las respuestas; una iteración futura podría incorporar evaluación automática de fidelidad (por ejemplo, *LLM-as-judge* o métricas tipo RAGAS).

Como mejora futura, el camino más natural sería ampliar el corpus a más temas del curso, experimentar con re-ranking de fragmentos y evaluar embeddings de mayor capacidad. Para el objetivo de esta actividad, la solución entregada documenta con claridad el problema, justifica el uso de `RAG + LLM` y la elección del modelo con evidencia comparativa, implementa un pipeline completo y reproducible (notebook, repositorio y demo en línea) y demuestra resultados verificables tanto cuantitativa como cualitativamente.
